docs/hugo/content/design/ADR-2024-10-OneOf-Resources/_index.md

```diff
@@ -0,0 +1,253 @@
+
+``` mermaid
+classDiagram
+
+class ClustersDatabase_Spec {
+    <<oneof>>
+    Name string
+}
+
+class ReadWriteDatabase
+class ReadOnlyFollowingDatabase
+
+ClustersDatabase_Spec --> ReadWriteDatabase : ReadWrite
+ClustersDatabase_Spec --> ReadOnlyFollowingDatabase : ReadOnlyFollowing
+```
+
+When the OneOf is rendered according to our current rules, we end up with this structure:
+
+``` mermaid
+classDiagram
+
+class ClustersDatabase_Spec {
+    <<oneof>>
+    ReadWrite ReadWriteDatabase
+    ReadOnlyFollowing ReadOnlyFollowingDatabase
+}
+
+class ReadWriteDatabase {
+    Name string
+}
+
+class ReadOnlyFollowingDatabase {
+    Name string
+}
+
+ClustersDatabase_Spec *-- ReadWriteDatabase
+ClustersDatabase_Spec *-- ReadOnlyFollowingDatabase
+```
+
+This is correct according to our rules for OneOf, but it doesn't work for ARM Spec generation due to the lack of a Name.
+
+#### Other factors
+
+We also have an `AzureName` property for many our custom resources, to allow the name in Azure to differ from the name in the cluster, due to different naming rules in each environment.
+
+## Option 1: Do Nothing
+
+Accept that the current decisions mean that we have to decline to support the `Microsoft.Kusto` resources, and any future resources that use this pattern.
+
+### Pros
+
+* Easy
+
+### Cons
+
+* Distasteful to decline to support a resource due to a technical limitation.
+* No guarantee that Kusto will be the only affected resource.
+* Risk that resources we already support will use this pattern in a new API version, rendering us unable to upgrade.
+
+## Option 2: Loosen the rules on Name
+
+For the specific case where the top level `spec` is a one-of, permit the `Name` to be omitted from the top level as long as it's present on all of the leaf types (guaranteeing that a name will always be present).
+
+### Pros
+
+* Avoids making already complex one-of handling more complex
+
+### Cons
+
+* May be confusing to users
+* Changes the existing rules, potentially breaking other code
+
+### Questions
+
+* How much of the existing operation of the code generator relies on the presence of `Name` at the top level on the ARM spec?
+* Ditto for the controller?
+
+## Option 3: Change the rules for all OneOf Types
+
+At the moment, the only permitted properties at the root level of a one-of are the mutually exclusive properties that represent the available options. We could loosen this rule to permit other properties, those in common to all leaf types, to be present at the root level.
+
+### Pros 
+
+* Conceptually simpler
+
+### Conts
+
+* Changing already complex one-of handling to make it more complex.
+* Requires changing the way we serialize/deserialize one-of types in non-trivial ways.
+* Potentially large blast-radius if our changes impact on one-of types we've already generated and released.
+
+## Option 4: Special case root OneOf Types
```

Comment:
There's a variant of this where we just generate two totally different resources for them... it wouldn't involve changing JSON marshalling, instead we'd just pretend it wasn't a OneOf and was two totally separate resources that look normal.

The problem would be... what are they called and how does that interact with ownership (because they're not two different "kinds" of owners as per ARM, it might cause issues for child resources).

Not sure it's a great idea because of those complexities but figured I'd mention it as it would probably be doable.